### PR TITLE
Go server command line options support

### PR DESCRIPTION
**Pull request by jessegross**
Support for command line options for controlling resource usage such as mlock, mmap and GPU allocation. In addition, switches support to more recent LoRA formats.

### CODE REVIEW

File: llama/llama.go [hunk 1]
```diff
@@ -205,17 +223,17 @@ func (m *Model) ShouldAddBOSToken() bool {
 	}
 }
 
-func (m *Model) ApplyLoraFromFile(loraPath string, scale float32, baseModelPath string, threads int) error {
+func (m *Model) ApplyLoraFromFile(context *Context, loraPath string, scale float32, threads int) error {
```
Comment (jmorganca) [inline llama/llama.go:226]: Did we want to remove the lora patch here or in a future change?

Comment (jessegross) [reply llama/llama.go:226]: Not totally sure I understand - loraPath is still needed in this version, it's just moved in the argument list. On the other hand, baseModelPath is no longer supported by the new versions of the functions and has been removed from the argument list.

File: llama/runner/runner.go [hunk 2]
```diff
@@ -659,22 +640,41 @@ func main() {
 	// otherwise Ollama can timeout for large model loads
 	// load the model
 	llama.BackendInit()
-	params := llama.NewModelParams(*nGpuLayers, *mainGpu, func(progress float32) {
-		slog.Debug("Loading model", "progress %", math.Round(float64(progress*100)))
-		server.progress = progress
-	})
+
+	var tensorSplitFloats []float32
```
[the remark this answers is not in the document]
Comment (jessegross) [reply llama/runner/runner.go:644]: There's already another variable named very similar to that, which is why I didn't call it that. I could rename the other one but agree that the names are a little ugly either way.

[the remark this answers is not in the document]
Comment (jmorganca) [reply llama/runner/runner.go:644]: sounds good

File: llama/llama.go (hunk 1 above)
Comment (jmorganca) [reply llama/llama.go:226]: Referring to this patch: https://github.com/ollama/ollama/blob/jmorganca/llama/llama/patches/09-lora.diff

Comment (jessegross) [reply llama/llama.go:226]: Ah, sorry, I thought "patch" was a typo for "path" given the quoted context.

My plan was to leave the patch for now and just not support it/use it in the Go server path. Once we remove server.cpp then we can remove the patch as well. Just trying to minimize collateral changes in the alternate server as we transition.

Comment (jessegross) [reply llama/llama.go:226]: That's a good point - I updated this commit to remove the LoRA patch from llama/patches

**Pull request merged**